### ISSUE TITLE
Restore from a remote Storage

### ISSUE THREAD

**Issue by dorianamouroux** [django-dbbackup/django-dbbackup#240]
I'm storing my backup onto Azure Blob Service. I'm using a custom remote storage to store my backup files. I'm using mysql database.
When I try to restore any file, I get this error : 
```
Finding latest backup
Restoring backup for database 'default' and server 'None'
Restoring: default-455a4e1caea4-2017-05-03-060045.dump
Restore tempfile created: 172.3 MiB
Are you sure you want to continue? [Y/n]
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/dbbackup/db/base.py", line 145, in run_command
    env=full_env)
  File "/usr/local/lib/python3.6/subprocess.py", line 665, in __init__
    errread, errwrite) = self._get_handles(stdin, stdout, stderr)
  File "/usr/local/lib/python3.6/subprocess.py", line 1162, in _get_handles
    p2cread = stdin.fileno()
io.UnsupportedOperation: fileno

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "manage.py", line 10, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/__init__.py", line 367, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python3.6/site-packages/django/core/management/__init__.py", line 359, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/base.py", line 294, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/base.py", line 345, in execute
    output = self.handle(*args, **options)
  File "/usr/local/lib/python3.6/site-packages/dbbackup/management/commands/dbrestore.py", line 54, in handle
    self._restore_backup()
  File "/usr/local/lib/python3.6/site-packages/dbbackup/management/commands/dbrestore.py", line 94, in _restore_backup
    self.connector.restore_dump(input_file)
  File "/usr/local/lib/python3.6/site-packages/dbbackup/db/base.py", line 96, in restore_dump
    result = self._restore_dump(dump)
  File "/usr/local/lib/python3.6/site-packages/dbbackup/db/mysql.py", line 39, in _restore_dump
    stdout, stderr = self.run_command(cmd, stdin=dump, env=self.restore_env)
  File "/usr/local/lib/python3.6/site-packages/dbbackup/db/base.py", line 156, in run_command
    "Error running: {}\n{}".format(command, str(err)))
dbbackup.db.exceptions.CommandConnectorError: Error running:  mysql dbname --host=db --port=3306 --user=******* --password=******
fileno
```

My function to download the file in my storage backend is something like this : 
```python
from azure.storage.blob import BlockBlobService
from django.core.files.base import ContentFile

def _open(self, name, mode='rb'):
    service = BlockBlobService(
        account_name=account_name,
        account_key=account_key)
    
    blob = service.get_blob_to_bytes(
        container_name="backups",
        blob_name=name)

    return ContentFile(blob.content)
```

It returns a Django Content File (https://docs.djangoproject.com/en/1.11/ref/files/file/#the-contentfile-class).
It's this instance of file that is injected in the `stdin` of `run_command` in `dbbackup/db/base.py`.

**Comment by ZuluPro** [comment]
Do you really think this issue is Djanog-DBBackup related ?
Otherwise, you should use [`get_blob_to_file`](https://azure-sdk-for-python.readthedocs.io/en/v0.11.1/ref/azure.storage.blobservice.html#azure.storage.blobservice.BlobService.get_blob_to_file) instead of `get_blob_to_bytes`.

**Comment by dorianamouroux** [comment]
The `_open` function should return a Python File Object.
I return a ContentFile, which is a in memory File object (https://github.com/django/django/blob/master/django/core/files/base.py#L134). It extends the Django File wrapper which extends the native Python File object.
It works just fine with user uploads and collect static. 

The function to restore should work with that.
Maybe if we use Python PIPE and `communicate` method, it would work

**Comment by ZuluPro** [comment]
Could you try the following things:
- Could you change this line :https://github.com/django-dbbackup/django-dbbackup/blob/master/dbbackup/db/base.py#L155 to raise the real error ?
- What if you try to laucnh the same thing than DBBackup manually ? (You can download you backup with DBBackup command and use the MySQL command provided by DBBbackup's traceback)

**Comment by dorianamouroux** [comment]
Here's the full stack trace if I raise the error : 

```
Traceback (most recent call last):
  File "/app/manage.py", line 10, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/__init__.py", line 367, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python3.6/site-packages/django/core/management/__init__.py", line 359, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/base.py", line 294, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/base.py", line 345, in execute
    output = self.handle(*args, **options)
  File "/app/apps/dbbackup/management/commands/dbrestore.py", line 54, in handle
    self._restore_backup()
  File "/app/apps/dbbackup/management/commands/dbrestore.py", line 94, in _restore_backup
    self.connector.restore_dump(input_file)
  File "/app/apps/dbbackup/db/base.py", line 96, in restore_dump
    result = self._restore_dump(dump)
  File "/app/apps/dbbackup/db/mysql.py", line 40, in _restore_dump
    stdout, stderr = self.run_command(cmd, stdin=dump, env=self.restore_env)
  File "/app/apps/dbbackup/db/base.py", line 155, in run_command
    raise err
  File "/app/apps/dbbackup/db/base.py", line 145, in run_command
    env=full_env)
  File "/usr/local/lib/python3.6/subprocess.py", line 665, in __init__
    errread, errwrite) = self._get_handles(stdin, stdout, stderr)
  File "/usr/local/lib/python3.6/subprocess.py", line 1162, in _get_handles
    p2cread = stdin.fileno()
io.UnsupportedOperation: fileno
```

You mean if I download the dump file, and I call myself something like `cat backup.dump | mysql <params>` ?

**Comment by dorianamouroux** [comment]
If I pipe my file this way : 
```
            process = Popen(cmd, stdin=PIPE, stdout=stdout, stderr=stderr,
                            env=full_env)
            process.communicate(input=stdin.read())
            process.wait()
```
It works well

**Comment by ZuluPro** [comment]
@dorianamouroux Thanks for your test !

> You mean if I download the dump file, and I call myself something like cat backup.dump | mysql <params> ?

Yes!

The problem seem to come from `ContentFile` object, it uses `io.StringIO` or `io.BytesIO` but theses classes doesn't support `fileno()` method. (There's no file description for theses in-memory files)

If you could provide a PR with your changes, there's a lot of work with because I think you have to fix unit tests too.
A short time solution could be to use `File` instead of `ContentFile`, I think Azure's objects support `fileno()`.

**Comment by dorianamouroux** [comment]
I created a quickfix for it https://github.com/django-dbbackup/django-dbbackup/pull/244

**Comment by jonathan-s** [comment]
Solved in #280
